[SPARK-18838][CORE] Introduce multiple queues in LiveListenerBus

bOOm-X opened the pull request:
## What changes were proposed in this pull request?

In this PR the single queue of the LiveListenerBus was replaced by multiple independent queues.
 - The EventLoggingListener was put in an independent queue (it was the most time consuming listener).
 - The executorAllocationManager was put in an independent qeue (it is very impacted by event drop of the main queue due to other listener impacts, so isolation purpose)
 - The UI listeners were put in the same independent queue (One of them was very time consuming, we want no impact of the GUI on the other listener)
 - The "extralisteners" were put in the same independent queue, for isolation purpose
 - The streamingListenerBus was put in an independent queue. this llistener is a bus too, and the processing time of its listeners can be quite significant


The queue and its processing thread have been extracted from `LiveListenerBus` into a class `BusQueue`. The definition of most of the methods of `ListenerBus` has been extracted into a trait `WithListenerBus`. The `LiveListenerBus` implements it directly. It hold  the "default" queue associated with a group of listeners and a list of queues. The method `addListener`  of the `WithListenerBus` has a new optional boolean parameter (default value false) to require a independent queue for this listener instead of the default one. This parameter is ignored in the default implementation (in`ListenerBus`) . 
A listener which is also a set of listeners has been added. It allows to keep the current behavior for a group of dependent listeners or the default queue. It handles the per listener metrics.
The methods `addProcessor` and `removeProcessor` have been added to `LiveListenerBus` to be able to add message processing at the super type `SparkListenerEvent` in addition to the per event type processing of the listener interface. 

## How was this patch tested?
utest + manual tests on the cluster

bOOm-X committed:
[SPARK-18838][CORE] Introduce multiple queues in LiveListenerBus

 The LiveListenerBus can now manage multiple queues for different listeners
 This will allow to increase a lot its deqeuing rate.
 All the listeners are still added to the main queue. So the behavior is the
 same as the previous one.
 In further commits some listeners will be moved to dedicated queues.

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

AmplabJenkins commented:
Can one of the admins verify this patch?

bOOm-X commented:
@vanzin @cloud-fan Can I have a review on this new PR ?

vanzin commented:
I'm busy, but I'll get to it eventually. You could at least write a proper commit summary in the meantime.

vanzin commented:
ok to test

SparkQA commented:
**[Test build #78001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78001/testReport)** for PR 18253 at commit [`fc6f609`](https://github.com/apache/spark/commit/fc6f6093763a43fae0e9474d801b9c9eb925326c).

SparkQA commented:
**[Test build #78001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78001/testReport)** for PR 18253 at commit [`fc6f609`](https://github.com/apache/spark/commit/fc6f6093763a43fae0e9474d801b9c9eb925326c).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78001/
Test FAILed.

bOOm-X commented:
Ok I rebased & updated my commit message.
I will update the PR message and push more commit to use what is done.

SparkQA commented:
**[Test build #78111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78111/testReport)** for PR 18253 at commit [`97cb911`](https://github.com/apache/spark/commit/97cb91104d4b6e560267d3d48a0a5215c92c8b14).

SparkQA commented:
**[Test build #78111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78111/testReport)** for PR 18253 at commit [`97cb911`](https://github.com/apache/spark/commit/97cb91104d4b6e560267d3d48a0a5215c92c8b14).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78111/
Test FAILed.

SparkQA commented:
**[Test build #78116 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78116/testReport)** for PR 18253 at commit [`18cb952`](https://github.com/apache/spark/commit/18cb952b6ab114524233e54a82a9095477153ecb).

SparkQA commented:
**[Test build #78116 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78116/testReport)** for PR 18253 at commit [`18cb952`](https://github.com/apache/spark/commit/18cb952b6ab114524233e54a82a9095477153ecb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78116/
Test FAILed.

SparkQA commented:
**[Test build #78117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78117/testReport)** for PR 18253 at commit [`52505d9`](https://github.com/apache/spark/commit/52505d95deba601d1b5cec2093c024eaa0da0af4).

SparkQA commented:
**[Test build #78117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78117/testReport)** for PR 18253 at commit [`52505d9`](https://github.com/apache/spark/commit/52505d95deba601d1b5cec2093c024eaa0da0af4).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78117/
Test FAILed.

vanzin commented:
I took a quick look and this does indeed look very much like work in progress. I also have a feeling that it's way over-engineered; there's a lot of base classes that are not that interesting, for example:

- why `SynchronousListenerBus`? It doesn't seem to add anything interesting over the existing interface.
- the whole "group of listeners" abstraction seems unnecessary. As far as I see it can be folded into the "listener queue" concept -  a queue feeds events to a collection of listeners.

Changing "post" to "postToAll" as part of this change also is adding a lot of unnecessary noise. I'm not a fan of the current class hierarchy of the listener bus and I think that change makes sense, but at the same time it should be done separately since it's distracting here.

I also saw methods that are not fully implemented in the code, so I assume you're still working on this.

I'd also like to see better justification for your custom queue implementation. Have you identified the use of `BlockingQueue` as a hotspot in the current code? My main worry is the sleep; if you're unlucky, your queue will always be 20ms behind in processing events and may suffer if there's a sudden burst while it's sleeping. You can probably squeeze better performance out of `BlockingQueue` without having to write your own - e.g. by using `drainTo` instead of reading events one by one.

The approach in #16291 had a lot of good things going for it, and mostly needed some clean up (and be modified to only change the live listener bus, and not the replay one). Your current approach seems a lot more complicated than that.

SparkQA commented:
**[Test build #78128 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78128/testReport)** for PR 18253 at commit [`a436b09`](https://github.com/apache/spark/commit/a436b091f657e8a019c21cb3a573dafccfd26a4c).

SparkQA commented:
**[Test build #78128 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78128/testReport)** for PR 18253 at commit [`a436b09`](https://github.com/apache/spark/commit/a436b091f657e8a019c21cb3a573dafccfd26a4c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78128/
Test FAILed.

SparkQA commented:
**[Test build #78134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78134/testReport)** for PR 18253 at commit [`9427be0`](https://github.com/apache/spark/commit/9427be042c8f817af08f1d8afe256b6ec708b38f).

SparkQA commented:
**[Test build #78134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78134/testReport)** for PR 18253 at commit [`9427be0`](https://github.com/apache/spark/commit/9427be042c8f817af08f1d8afe256b6ec708b38f).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78134/
Test FAILed.

SparkQA commented:
**[Test build #78136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78136/testReport)** for PR 18253 at commit [`d429be6`](https://github.com/apache/spark/commit/d429be6748e1588c27729436fc9ef20e2590900e).

SparkQA commented:
**[Test build #78136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78136/testReport)** for PR 18253 at commit [`d429be6`](https://github.com/apache/spark/commit/d429be6748e1588c27729436fc9ef20e2590900e).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78136/
Test FAILed.

bOOm-X committed:
[SPARK-18838][CORE] eventLoggingListener dedicated queue

The eventLoggingListener is now in a dedicated asynchronous queue.
This listener could represent 50% of the event processing time of the
standard queue

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

SparkQA commented:
**[Test build #78138 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78138/testReport)** for PR 18253 at commit [`1fe9161`](https://github.com/apache/spark/commit/1fe916198e3308418e1f7ced57e44173ab278e22).

bOOm-X committed:
[SPARK-18838][CORE] ExecutorAllocationManager dedicated queue

The ExecutorAllocationManager is now in a dedicated asynchronous queue.
This listener suffer a lot of event drops. Put it in a dedicated queue
decrease a lot the chance of them

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

SparkQA commented:
**[Test build #78139 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78139/testReport)** for PR 18253 at commit [`123aa92`](https://github.com/apache/spark/commit/123aa928f3492b7f582644e20c9262ca91ac779f).

SparkQA commented:
**[Test build #78138 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78138/testReport)** for PR 18253 at commit [`1fe9161`](https://github.com/apache/spark/commit/1fe916198e3308418e1f7ced57e44173ab278e22).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdated(blockUpdatedInfo: BlockUpdatedInfo)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78138/
Test FAILed.

SparkQA commented:
**[Test build #78139 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78139/testReport)** for PR 18253 at commit [`123aa92`](https://github.com/apache/spark/commit/123aa928f3492b7f582644e20c9262ca91ac779f).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78139/
Test FAILed.

bOOm-X committed:
[SPARK-18838][CORE] UI event listeners dedicated queue

The UI event listeners are now in a dedicated asynchronous queue.
This set of listener could represent 40% of the event processing time
+ do not block the listener bus with call from the GUI

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

SparkQA commented:
**[Test build #78141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78141/testReport)** for PR 18253 at commit [`b73b020`](https://github.com/apache/spark/commit/b73b020c9b21bbf07059bbef158cd0c64bb97853).

SparkQA commented:
**[Test build #78141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78141/testReport)** for PR 18253 at commit [`b73b020`](https://github.com/apache/spark/commit/b73b020c9b21bbf07059bbef158cd0c64bb97853).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78141/
Test FAILed.

bOOm-X committed:
fix tests

SparkQA commented:
**[Test build #78146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78146/testReport)** for PR 18253 at commit [`a239f83`](https://github.com/apache/spark/commit/a239f8340b54c04abf5fcae7c9ff673d3209e03b).

SparkQA commented:
**[Test build #78146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78146/testReport)** for PR 18253 at commit [`a239f83`](https://github.com/apache/spark/commit/a239f8340b54c04abf5fcae7c9ff673d3209e03b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78146/
Test FAILed.

SparkQA commented:
**[Test build #78165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78165/testReport)** for PR 18253 at commit [`24721b9`](https://github.com/apache/spark/commit/24721b946c5a1a3fbf1332caa3509eae3d2f354e).

SparkQA commented:
**[Test build #78165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78165/testReport)** for PR 18253 at commit [`24721b9`](https://github.com/apache/spark/commit/24721b946c5a1a3fbf1332caa3509eae3d2f354e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #78167 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78167/testReport)** for PR 18253 at commit [`cafcd96`](https://github.com/apache/spark/commit/cafcd96560c140fc6d9d4338977c03820f2cb5bf).

SparkQA commented:
**[Test build #78167 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78167/testReport)** for PR 18253 at commit [`cafcd96`](https://github.com/apache/spark/commit/cafcd96560c140fc6d9d4338977c03820f2cb5bf).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78167/
Test FAILed.

SparkQA commented:
**[Test build #78168 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78168/testReport)** for PR 18253 at commit [`13003b2`](https://github.com/apache/spark/commit/13003b21f90d9254a29222094efa5bede2bd98ab).

SparkQA commented:
**[Test build #78169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78169/testReport)** for PR 18253 at commit [`d3d2cbe`](https://github.com/apache/spark/commit/d3d2cbe32594d3dd26432dd9495c2a5f3242c78b).

SparkQA commented:
**[Test build #78168 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78168/testReport)** for PR 18253 at commit [`13003b2`](https://github.com/apache/spark/commit/13003b21f90d9254a29222094efa5bede2bd98ab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78168/
Test FAILed.

SparkQA commented:
**[Test build #78169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78169/testReport)** for PR 18253 at commit [`d3d2cbe`](https://github.com/apache/spark/commit/d3d2cbe32594d3dd26432dd9495c2a5f3242c78b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78169/
Test FAILed.

SparkQA commented:
**[Test build #78186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78186/testReport)** for PR 18253 at commit [`be3d560`](https://github.com/apache/spark/commit/be3d5604970756a14b7d15ea086e670e630555d5).

SparkQA commented:
**[Test build #78186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78186/testReport)** for PR 18253 at commit [`be3d560`](https://github.com/apache/spark/commit/be3d5604970756a14b7d15ea086e670e630555d5).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78186/
Test FAILed.

SparkQA commented:
**[Test build #78205 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78205/testReport)** for PR 18253 at commit [`7a5df2d`](https://github.com/apache/spark/commit/7a5df2d9f39e5c52a5e8111f66a7e33e98523d68).

SparkQA commented:
**[Test build #78205 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78205/testReport)** for PR 18253 at commit [`7a5df2d`](https://github.com/apache/spark/commit/7a5df2d9f39e5c52a5e8111f66a7e33e98523d68).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78205/
Test FAILed.

SparkQA commented:
**[Test build #78219 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78219/testReport)** for PR 18253 at commit [`4596c61`](https://github.com/apache/spark/commit/4596c616d4508d8a10955d461167b2674fb3d644).

SparkQA commented:
**[Test build #78219 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78219/testReport)** for PR 18253 at commit [`4596c61`](https://github.com/apache/spark/commit/4596c616d4508d8a10955d461167b2674fb3d644).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78219/
Test FAILed.

SparkQA commented:
**[Test build #78225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78225/testReport)** for PR 18253 at commit [`965b105`](https://github.com/apache/spark/commit/965b105ea64173f70bb54af1eff73389354c7f20).

SparkQA commented:
**[Test build #78225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78225/testReport)** for PR 18253 at commit [`965b105`](https://github.com/apache/spark/commit/965b105ea64173f70bb54af1eff73389354c7f20).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78225/
Test PASSed.

bOOm-X committed:
[SPARK-18838][CORE] extralisteners dedicated queue

The extralisteners are now in a dedicated asynchronous queue.
So they cannot interfere with the execution of the spark internal listeners

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

bOOm-X committed:
[SPARK-18838][CORE] streaming listener dedicated queue

The streaming listener which is a bus too (for streaming event & listener)
is now in a dedicated asynchronous queue.
So they streaming listeners are run without impact from the other listeners

  ## How was this patch tested?
  unit test  + manual tests have been run on the cluster

SparkQA commented:
**[Test build #78226 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78226/testReport)** for PR 18253 at commit [`16ba70a`](https://github.com/apache/spark/commit/16ba70ac6f8d3fe51c4a912fae1f7cb45f27109c).

SparkQA commented:
**[Test build #78226 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78226/testReport)** for PR 18253 at commit [`16ba70a`](https://github.com/apache/spark/commit/16ba70ac6f8d3fe51c4a912fae1f7cb45f27109c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78226/
Test PASSed.

bOOm-X commented:
@vanzin Ok it is ready now.

> why SynchronousListenerBus? It doesn't seem to add anything interesting over the existing interface.

I removed it.

> the whole "group of listeners" abstraction seems unnecessary. As far as I see it can be folded into the "listener queue" concept - a queue feeds events to a collection of listeners.

I simplified it and add usage in the other commits. It is basically usefull to hold the metrics, and I need a common way to add a group of dependent listeners to the LiveListenerBus and the ReplayBus.

> Changing "post" to "postToAll" as part of this change also is adding a lot of unnecessary noise. I'm not a fan of the current class hierarchy of the listener bus and I think that change makes sense, but at the same time it should be done separately since it's distracting here

100% agree. I removed it

> I'd also like to see better justification for your custom queue implementation. Have you identified the use of BlockingQueue as a hotspot in the current code?

This implementation had 2 advantages: it is a 1 producer - 1 consumer queue whereas the BlockingQueue is a n producers - m consumers. So it use much less synchronization. The other advantage (the main one) is that no object is created for each message added to the queue. So it produces a lot less garbage. More independent queues we have, more it is significant.

> My main worry is the sleep; if you're unlucky, your queue will always be 20ms behind in processing events and may suffer if there's a sudden burst while it's sleeping. 

I change it to 1 ms instead of 20ms. This time is much less than the average processing time of the fastest listener (around 5 ms for the  HeartbeatListener). It is just to force the consumer thread to escape in case of empty queue to give more chance to the producer thread to be scheduled. I can remove it if you want.

vanzin commented:
> This implementation had 2 advantages... it use much less synchronization.

Yes, but have you quantified how much you win with that? If the blocking queue approach has enough throughput for the listener bus, it's safer to use it.

> is that no object is created for each message added to the queue

Well, you could use an `ArrayBlockingQueue`. Then no extra object is allocated either.

Here's a link with numbers for `ArrayBlockingQueue`:
https://github.com/LMAX-Exchange/disruptor/wiki/Performance-Results

4M ops per sec in the 1P-1C case looks plenty fast for Spark's need.

> I change it to 1 ms instead of 20ms. 

I think if you really insist on going this route, you should use `LockSupport.park/unpark` instead of fighting for CPU time like this.

bOOm-X committed:
[SPARK-18838][CORE] fix

 - wait on empty queue instead of looping

SparkQA commented:
**[Test build #78412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78412/testReport)** for PR 18253 at commit [`7c4c040`](https://github.com/apache/spark/commit/7c4c04022aabda2f1b056b4ff1e3bc8f0f1aa5cf).

bOOm-X commented:
> Well, you could use an ArrayBlockingQueue. Then no extra object is allocated either.

Yes I agree. But you get the synchronization too. I am still agree that it should not have a big impact yet. But using an ArrayBlockingQueue does not simplify the code a lot. The current implementation is not complicated, not too verbose, and base on a simple pure scala array. I do not think that it has a huge complexity cost compared to the java ArrayBlockingQueue.

I change the Thread.sleep to a Thread.yield to be less agressive for the thread unscheduling. Even with very few messages it should not consume too much CPU, and it will be much more reactive when messages are bursting.

SparkQA commented:
**[Test build #78412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78412/testReport)** for PR 18253 at commit [`7c4c040`](https://github.com/apache/spark/commit/7c4c04022aabda2f1b056b4ff1e3bc8f0f1aa5cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78412/
Test PASSed.

vanzin commented:
> I change the Thread.sleep to a Thread.yield to be less agressive for the thread unscheduling

How does that help? Now you have a thread that never sleeps. It gives up its scheduler slot but is still active and will be scheduled again, even if it doesn't need to.

Can you explain in numbers why exactly you're against any of the other solutions that actually prevent these threads from being scheduled when not needed?

vanzin commented:
To illustrate what I'm trying to say, here's what `top` says for an *idle* spark-shell with your code:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                             
24247 vanzin    20   0 6778304 766336  31836 S 201.0  2.3   1:34.06 java  
```

I hope you see why that's an issue.

bOOm-X commented:
With a `LockSupport.parkNanos(1000)` instead I get:

```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
43448 a.prang   20   0 17.663g 2.378g  66668 S  35.2  0.9   3:41.04 java
```
Is it ok ?

vanzin commented:
And idle spark-shell should be using 0% of the CPU.

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                             
 1304 vanzin    20   0 6773996 743104  31468 S   0.3  2.3   0:37.28 java
```

SparkQA commented:
**[Test build #78569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78569/testReport)** for PR 18253 at commit [`f423363`](https://github.com/apache/spark/commit/f423363d2712cfec7fd93f5ff2ef1a078408ce9f).

SparkQA commented:
**[Test build #78569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78569/testReport)** for PR 18253 at commit [`f423363`](https://github.com/apache/spark/commit/f423363d2712cfec7fd93f5ff2ef1a078408ce9f).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78569/
Test FAILed.

bOOm-X commented:
please test

SparkQA commented:
**[Test build #78624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78624/testReport)** for PR 18253 at commit [`a3b193a`](https://github.com/apache/spark/commit/a3b193a5511be12282db020cd48d5518d86da696).

SparkQA commented:
**[Test build #78624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78624/testReport)** for PR 18253 at commit [`a3b193a`](https://github.com/apache/spark/commit/a3b193a5511be12282db020cd48d5518d86da696).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78624/
Test FAILed.

bOOm-X commented:
`PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
11312 a.prang   20   0 17.661g 1.699g  66924 S   8.3  0.7   1:53.41 java`

Ok it is good now.
Do you have an idea why the PySpark packaging test are failling ?
I tried to rebase but no impact

bOOm-X commented:
@vanzin : Any news ?

cloud-fan commented:
is this a successor of https://github.com/apache/spark/pull/16291 ?

bOOm-X commented:
@cloud-fan Yes it is !

bOOm-X commented:
retest this please

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #79271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79271/testReport)** for PR 18253 at commit [`a3b193a`](https://github.com/apache/spark/commit/a3b193a5511be12282db020cd48d5518d86da696).

SparkQA commented:
**[Test build #79271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79271/testReport)** for PR 18253 at commit [`a3b193a`](https://github.com/apache/spark/commit/a3b193a5511be12282db020cd48d5518d86da696).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79271/
Test FAILed.

SparkQA commented:
**[Test build #79296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79296/testReport)** for PR 18253 at commit [`19ee415`](https://github.com/apache/spark/commit/19ee415f3a074e97d6f550fed46a8556ef3ce62b).

SparkQA commented:
**[Test build #79296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79296/testReport)** for PR 18253 at commit [`19ee415`](https://github.com/apache/spark/commit/19ee415f3a074e97d6f550fed46a8556ef3ce62b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79296/
Test PASSed.

cloud-fan commented:
What's the processing model now? Previously it's very simple, we have an event queue, and many threads send events to this queue, and one thread poll events from this and call listeners. What is it now?

Can you give a summary about the implementation? e.g. How you change the API in `ListenerBus` and what's the new class hierarchy for listener buses?

vanzin commented:
I was out for the last couple of weeks. But, taking a quick look, I really would prefer for you to stop trying to implement your own blocking queue without numbers that justify the need for it. Putting event listeners in separate queues does not require you to write a custom blocking queue implementation; if there's benefit in that, you can do it as a separate PR, when you actually provide evidence that backs up the necessity for it.

SparkQA commented:
**[Test build #79517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79517/testReport)** for PR 18253 at commit [`4edcee7`](https://github.com/apache/spark/commit/4edcee7f68b95dd78a6a487aa054cbd9bafcae66).

SparkQA commented:
**[Test build #79517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79517/testReport)** for PR 18253 at commit [`4edcee7`](https://github.com/apache/spark/commit/4edcee7f68b95dd78a6a487aa054cbd9bafcae66).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79517/
Test FAILed.

Commit:
to ArrayBloquingQueue

SparkQA commented:
**[Test build #79524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79524/testReport)** for PR 18253 at commit [`aca4dc5`](https://github.com/apache/spark/commit/aca4dc5b440576d8a35e4b2c083eb217dbeaaaf2).

SparkQA commented:
**[Test build #79524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79524/testReport)** for PR 18253 at commit [`aca4dc5`](https://github.com/apache/spark/commit/aca4dc5b440576d8a35e4b2c083eb217dbeaaaf2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79524/
Test PASSed.

bOOm-X commented:
@vanzin I put an ArrayBlockingQueue as you wanted

bOOm-X commented:
@cloud-fan 
There are now multiple queues. Many threads send events to a dispatcher which copy the event in each queue. One thread per queue poll the events from its dedicated queue and call its attached listener(s).

For listener bus there is now two pure traits. the first one WithListenerBus allow to add / remove listener (current behavior). By default all the current implementation implement this trait . Another trait WithMultipleListenerBus inherit from the previous one and add a method to add isolated listener. Only the LiveListenerBus and the ReplayListenerBus implement this trait. For the second one, it is just a synonym of the simple addListener method. All the UI listeners are added to both through this method as a group of listeners (they are inter-dependent).
The StreamingListenerBus, the EventLoggingListener, the ExecutorAllocationManager are added to the LiveListenerBus (with this method) to a dedicated queue. The "extralisteners" are added to it as a group of listeners in a dedicated queue too. All other listeners are added to the "default" queue with the old method.

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -532,7 +533,10 @@ class SparkContext(config: SparkConf) extends Logging {
             _conf, _hadoopConfiguration)
         logger.start()
-        listenerBus.addListener(logger)
+        listenerBus.addProcessor(
```
I'm having a hard time finding the declaration of this method. I can't find it in your code nor in the existing master branch. Can you link to it?

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -2350,13 +2354,12 @@ class SparkContext(config: SparkConf) extends Logging {
         conf.get("spark.extraListeners", "").split(',').map(_.trim).filter(_ != "")
-      for (className <- listenerClassNames) {
-        // Use reflection to find the right constructor
+      val extraListeners = listenerClassNames.map{ className =>
```
`.map { className =>`

You have a lot of style issues in your code - indentation, spacing, etc. Please read the style section in http://spark.apache.org/contributing.html and try to follow it.

vanzin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -2379,8 +2382,13 @@ class SparkContext(config: SparkConf) extends Logging {
+        listener
+      }
+      if (extraListeners.nonEmpty) {
+        val group = new FixGroupOfListener(extraListeners, "extraListeners")
```
`FixGroupOfListener` is a bad class name. I'm not even sure what it's supposed to be, but the closest I can think of is `ListenerGroup`.

But perhaps this shouldn't be exposed at all. If you add a queue name to the listener registration method, you can hide this from callers altogether. That is, if I understood what this class is in the first place.

Then you wouldn't need `addIsolatedListener` either.

vanzin replied:
> First, I think that modifying the existing addListener method is a bad idea. It will impact a lot the code.

That's why overloaded methods exist.

> But I have 2 kind of group of listeners

I don't think there's really a distinction between the two types of groups you mention. The "UI group" is just a modifiable group that you don't modify after it's been created.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -227,6 +169,7 @@ private[spark] class EventLoggingListener(
    * ".inprogress" suffix.
    */
   def stop(): Unit = {
+    flush()
```
Shouldn't be necessary (`close()` does it).

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -131,94 +132,35 @@ private[spark] class EventLoggingListener(
-
-  override def onOtherEvent(event: SparkListenerEvent): Unit = {
+  def log(event: SparkListenerEvent): Unit = {
     if (event.logEvent) {
```
Since you're adding an event filter, you could perform this check there...

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/bus/ListenerBusQueueImpl.scala:
```diff
@@ -0,0 +1,122 @@
+import org.apache.spark.scheduler.bus.ListenerBusQueue.{FixGroupOfListener, ModifiableGroupOfListener}
+
+// For generic message processor (like event logging)
+private[scheduler] class ProcessorListenerBusQueue(
```
First, the name of this file is weird. But more importantly, why are these classes even necessary?

Why can't you have a single queue implementation that manages a group of listeners? Whether the group has a single listener or multiple shouldn't matter - the implementation can be the same.

bOOm-X replied:
For the name of the file, I can change it ! Do you have a better name ? I can even put the content of the file (the 2 concrete implementations)  in the BusQueue.scala file.

I refactored a bit this file. Now I have only 2 implementations: 
 - `ProcessorBusQueue`: This is the implementation for generic processor (In which we do not do the dispatch by  event type)
-  `ListenerBusQueue`: This is the implementation for listener (with the dispatch by event type)

vanzin replied:
I'm still confused about why you need 2 implementations. Why doesn't `ListenerBusQueue` work for everybody? And why shouldn't it?

I need more time to actually grok all this code, but like Wenchen suggested before, this is a big change and it would benefit from a more detailed explanation of exactly how you're organizing the hierarchy of listener, groups, etc. Your PR description only explains which queues you created, but not any of the changes that were needed to achieve that.

If it makes it easier, you can create a `README.md` file with a longer explanation for how things are organized. (for example, check `ommon/network-common/src/main/java/org/apache/spark/network/crypto` where I added a README to explain details of what that whole body of code is doing).

vanzin reviewed (commented):
I started reviewing, but again, I noticed the same thing I commented on before. This is way over-engineered. You can do this in a much, much simpler way. There's no need to create all the different abstractions you're adding - the current listener abstraction is enough to achieve what is being proposed here. 

All you need is to add a "queue name" parameter to the `addListener` method, and potentially an "event filter" parameter. Everything else is hidden in the listener implementation, and doesn't need to be exposed to any calling code.

AmplabJenkins commented:
Can one of the admins verify this patch?

abellina commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -139,30 +148,60 @@ private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
+      started = true
     } else {
-      onDropEvent(event)
+      throw new IllegalStateException("LiveListener bus already started!")
```
definitely want to unlock before this.

abellina commented on core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala:
```diff
@@ -27,7 +27,12 @@ private[spark] trait SparkListenerBus
   protected override def doPostEvent(
       listener: SparkListenerInterface,
-      event: SparkListenerEvent): Unit = {
+      event: SparkListenerEvent): Unit = SparkListenerEventDispatcher.dispatch(listener, event)
```
why is this change necessary?

abellina reviewed (commented):
Hi just some comments as I try to understand the code.

Commit:
rebase and review

bOOm-X commented:
retest this please

bOOm-X committed:
simplify

remove WithMultipleListenerBus and replace it by adding a boolean

bOOm-X commented:
@vanzin I simplify a lot the code. There are now only one implementation for the queue and for the group of listeners. I removed the extra trait in the listener hierarchy too.
Can you take a look ?

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -39,98 +40,114 @@ import org.apache.spark.util.Utils
+   /**
+    * Add a generic listener to an isolated pool.
+    */
+  def addProcessor(processor: SparkListenerEvent => Unit,
```
what's the difference between `addProcessor` and `addListener`?

bOOm-X replied:
with addProcessor, you do not have to provide a SparkListenerInterface (So with a method per message type), but just a generic function which handle  SparkListenerEvent  (the super type of each event type). So when you do a generic processing (see EventLoggingListener for example)  it is very convenient, and cherry on the top you avoid the horrible and costly dispatch function , which in this case (generic processing) is a burden

cloud-fan replied:
shall we do it in a separated PR?

bOOm-X replied:
It is just a small technical refactoring which came for almost free with the new qeue object. It is also very convenient to be able to handle the asynchronous LiveListenerBus in the test.  I think that we can keep it in this PR.

cloud-fan commented:
The PR description is good for explaining the new behavior, but can you say more about the implementation?

IMO, we just need to duplicate the event queue for each important listener like event logging listener, and non-important listeners can share one event queue like the current behavior. Then each event queue is processed by an individual thread.

bOOm-X commented:
@cloud-fan I will update the PR description with some details on the implementation

cloud-fan commented:
I think `BusQueue` is a good abstraction, but we can still simplify other parts. My proposal:
```
// Do we need a better name like ListenesGroup? it's very similar to the previous LiveListenerBus.
class BusQueue extends SparkListenerBus {
  val eventQueue = ...
}

class LiveListenerBus {
  val listenersGroups: Map[String, BusQueue] = ...

  def addListener(listener, groupName = "default") {
    val group = listenerGroups.getOrUpdate(groupName, new BusQueue ...)
    group.addListener(listener)
  }

  def post(event) {
    listenersGroups.foreach { case (_, group) =>
      group.post(event)
    }
  }
}

```

vanzin commented:
Agree that it still seems like there's too many moving parts here. I don't see a whole lot of difference between `BusQueue` and `GroupOfListener` (which is a weird name, btw). A queue can have a bunch of listeners and a `LiveListenerBus` can have a bunch of queues. I don't see the need for anything more complicated than that.

`WithListenerBus` is another thing I don't understand; aside from the really weird name, it seems to be mostly the same thing as `ListenerBus`.

Also to reinforce a previous comment, your code has a ton of style issues. It doesn't matter that checkstyle doesn't complain about them; you still have to follow the code convention of the project or we'll be forever pointing out style issues in your code.

bOOm-X commented:
@cloud-fan 
In my opinion, having the queues indexed by a string and reflecting that in the API is a bit too error prone in this case. What you want is to be in an isolated queue. With this kind of API it is easy to have an other listener added to the same queue because of a conflicting label. In the contrary, it is quite easy to not had 2 depending listeners not in the same queue because of case, ..., of the label. With a map of queues, it enforces the mutability of the queue, which is not really necessary here. And I think that it is a good idea to enforce the fact that depending listeners should be put in the same queue, it increases the readability and the clarity of the API.

bOOm-X commented:
@vanzin 
`GroupOfListener` is just a set of listeners, which handle all the metric stuff. it allows to decouple this metric aspect from the `ListernerBus` and the `LiveListenerBus`. And It allows to enforce the dependency model between listeners (one message processed sequentially). I can change its name if you want. 
The `BusQueue` just handle the queuing / dequeueing generic processing, which includes the consumer thread start an stop and the queuing strategy (drop if full).

`WithListenerBus` is indeed just the declaration of the method implemented in `ListenerBus`. This declaration is shared between the `LiveListenerBus` and the `ReplayListenerBus`, and the UI listeners are added to them through it. The `ListenerBus` still contains a simple synchronous implementation of them used by the `ReplayListenerBus`. The `LiveListenerBus` has its own based on multiple queues for some kind of set of listeners. I can rename this interface if you want.

I will do a pass to try to fix the code style issues.

vanzin commented:
I really dislike `WithListenerBus` - both as a name and as a concept. There's already a `ListenerBus` trait; if it's not enough or is broken in some way, it should be fixed, instead of being patched by introducing yet more complexity in the hierarchy.

I think part of the confusion here is that the current code is trying to both refactor the `ListenerBus` hierarchy and add the concept of queues are the same time. From my point of view they're different things and could be done separately. For example you could add queues to `LiveListenerBus` only, which is really the only place that matters in the end. Maybe it won't be optimal in its first iteration, but it would be a much easier change to review.

I don't doubt that there's benefit in taking a holistic look into this part of the class hierarchy; but it would be good to do that separately, both so that we can clearly see that the proposed hierarchy makes sense, and so that it's easier to review things. It's easier to wrap your head around the code if it's focused on one problem instead of two.

vanzin commented:
@bOOm-X 

I pushed some code to my repo: https://github.com/vanzin/spark/tree/SPARK-18838

Which is an attempt to do things the way I've been trying to explain. It tries to keep changes as local as possible to `LiveListenerBus`, creating a few types that implement the grouping and asynchronous behavior. You could do filtering by extending the new `AsyncListener`, for example, and adding it to the live listener bus.

It's just a p.o.c. so I cut a few corners (like metrics), and I only ran `SparkListenerSuite`, but I'm just trying to show a different approach that leaves the `ListenerBus` hierarchy mostly the same as now.

bOOm-X commented:
@vanzin I pushed some comments on your code. I think that trying to keep the exact same class hierarchy leads to a very complex code, with many drawbacks.

vanzin commented:
You commented on my code, not on the idea. My code was hacked together quickly, it can be cleaned up a lot. Your comments don't disprove that separating the refactoring of the listener bus hierarchy from the introduction of queues is impossible or undesirable.
